kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/PaginatedScanFilesIteratorImpl.java

```diff
@@ -34,21 +46,68 @@ public PaginatedScanFilesIteratorImpl(
     this.pageSize = paginationContext.getPageSize();
   }
 
+  // Bug: we can't call hasNext() twice
   @Override
-  public Row getCurrentPageToken() {
-    return null;
+  public boolean hasNext() {
+    if (closed) {
+      throw new IllegalStateException("Can't call `hasNext` on a closed iterator.");
+    }
+    if (!nextBatch.isPresent()) {
+      prepareNext();
+    }
+    return nextBatch.isPresent();
   }
 
-  @Override
-  public boolean hasNext() {
-    return false;
+  private void prepareNext() {
+    if (numScanFilesReturned >= pageSize) {
+      return;
+    }
+    if (originalIterator.hasNext()) {
```

Comment:
tip: reduce and eliminate indentation when you can: you could do `if (!originalIterator.hasNext()) { return; }` and then all the code within the brackets can be shifted left (no identation). this makes the code more readable